src/ImageSharp/Formats/Png/Zlib/Adler32.cs

```diff
@@ -4,145 +4,283 @@
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    length -= 16;
+                }
 
-            s1 = (s1 + ((uint)value & 0xFF)) % Base;
-            s2 = (s1 + s2) % Base;
+                while (length-- > 0)
+                {
+                    s2 += s1 += Unsafe.Add(ref bufferRef, index++);
+                }
+
+                if (s1 >= BASE)
+                {
+                    s1 -= BASE;
+                }
 
-            this.checksum = (s2 << 16) + s1;
+                s2 %= BASE;
+            }
+
+            return s1 | (s2 << 16);
         }
+#endif
 
-        /// <inheritdoc/>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void Update(ReadOnlySpan<byte> data)
+        [MethodImpl(InliningOptions.HotPath | InliningOptions.ShortMethod)]
+        private static uint CalculateScalar(uint adler, ReadOnlySpan<byte> buffer)
         {
-            ref byte dataRef = ref MemoryMarshal.GetReference(data);
-            uint s1 = this.checksum & 0xFFFF;
-            uint s2 = this.checksum >> 16;
+            uint s1 = adler & 0xFFFF;
+            uint s2 = (adler >> 16) & 0xFFFF;
+            uint k;
 
-            int count = data.Length;
-            int offset = 0;
+            ref byte bufferRef = ref MemoryMarshal.GetReference<byte>(buffer);
+            uint length = (uint)buffer.Length;
+            int index = 0;
 
-            while (count > 0)
+            while (length > 0)
             {
-                // We can defer the modulo operation:
-                // s1 maximally grows from 65521 to 65521 + 255 * 3800
-                // s2 maximally grows by 3800 * median(s1) = 2090079800 < 2^31
-                int n = 3800;
-                if (n > count)
+                k = length < NMAX ? length : NMAX;
+                length -= k;
+
+                while (k >= 16)
                 {
-                    n = count;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
+                    s1 += Unsafe.Add(ref bufferRef, index++);
+                    s2 += s1;
```

Comment:
JIT doesn't do great with code like this.

On legacy jit32, it emits this: 

```asm
    L0031: movzx eax, byte [ebx+eax]
    L0035: add edx, eax
    L0037: add ecx, edx
    L0039: mov eax, esi
    L003b: inc esi
    L003c: movzx eax, byte [ebx+eax]
    L0040: add edx, eax
    L0042: add ecx, edx
    L0044: mov eax, esi
    L0046: inc esi
    …
```

RyuJIT does a bit better by not *actually* incrementing `index` each time, but it emits an extra `mov` for each block because it's confused by your two variables:

```asm
    L0031: movzx r9d, byte ptr [rax+1]
    L0036: add r9d, ecx
    L0039: mov ecx, r9d
    L003c: add r8d, ecx
    L003f: movzx r9d, byte ptr [rax+2]
    L0044: add r9d, ecx
    L0047: mov ecx, r9d
    L004a: add r8d, ecx
    …
```

Writing it more like the C code:

```C#
    byte* pbuff = pbuffer + index;
    if (length >= 16) 
    { 
        s2 += (s1 += pbuff[0]); 
        s2 += (s1 += pbuff[1]); 
        s2 += (s1 += pbuff[2]); 
        ...
    }
```

gives better codegen in both:

```asm
    L0042: movzx r9d, byte ptr [rcx+1]
    L0047: add eax, r9d
    L004a: add r8d, eax
    L004d: movzx r9d, byte ptr [rcx+2]
    L0052: add eax, r9d
    L0055: add r8d, eax
```

[SharpLab for both variants](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgyEMA3jQYem9jrx4MAsgAUsmLYXAA8wACeGDAAfAzAHABmyTIAlO6ebtSeeQywyYkxMIkpaVBsMEUAvAXVATD40FH+2FC4ABbYkiwA4jAYVRUwXCqBSakZagw2ufkevhgM0lwA5hidDHWTFSwAMqMbnRrzC1kLAPSXDJuqpQDuUpIM3Qqlze8AJgzJUKK3TqlWQggCi9kkEFMszO+R8flw5G2DHIpwWHnhy1wjjqqIu+SWDF8XxgCGRdDR6PxeV4RUCq2ODDiuKQ6Rh6Nc7I5nkRDAA1HUAKpcXDYNIsACCXy+gUKZSmlWqaCJXBJCD5fPSM2p6Ox/LqiO1sI5vIFDGFovFUplct2MmGyuJpI1Wq53IYerNhrd3NNQpFYsE1tlDTtiuSjtVzs1RvdHk9BtRPpNSLNFsDkulIaKYYdKrVLtjcYTHqTOoWfvNAatWdt5XtSvz0dd5fyJe9rbylfTNZtofr4cjBZjyd1ji9ZeNutT/stQdr/YVead6pHnZ548TRfd3er877OYHy6jq5bU4rm9L299M6rc8zB/lFWPw7PcY9l4757bt57++zT4NhGTanteJqfpO767vewZ1kujYroWo4XvqV7IT+qF/g+AG5ghJ5Ieu8YQWB06YXu2Fwc+eGviRKETrRGFpuRsGLlRwGIWu35dsR6Fdr+zELoe8HsfhnFQTxhGlmRMGCYBg4gQRXEbqhX7vgymwMHALIMQAvjCrYPJ0UilPSRybHAcBMgwdBsq2Li8cpE7SRmLFCWxQ7Nrp6GtsQADsUmEAwgR6uE4QoqylIeDpcx5MQ3h7gwmIMAAcsEoQRNEsQJGGmSwjkHLJLwCAwD8EwlAAVAwYhhsiOWuDF6KEupWw7AOBxmScrattcgL3AwTySC8bwfBA3y/P8+C9QwIKyOCKxQmI6FJbyuKRXChJ6qtraEiu5JrRyraZTAlXVeUyKnQq/Igftni0kFzVMiytnfvZkklsFt4XckADadAALpvsWl4fahX3feQAMMdxqEg2aYOkJDDlETDlZg7YiNvcDqO7N9KAY0pyNmrDdRgwArPj4ko59ONIBTQNU6DOO+XTO5Y9T5TfQAHCzvps4zHMAJw8+BDNwzj5D/YDrOiyT4sQ1LvMy1V4sIwrItE9jHPkOjatjkrYPkHjut0XUxPK1r5PG/kX2oeQSA3QsD1aeFXkNQshnGfdHUWVZNnoa9BMfkr5VfYpHJ6V135+QFQUhWFdtaq20WwsnOlAA===)

It's also not clear that unroll by 16 is ideal for modern processors.  This may be an outdated optimization from zlib.  Might be worth testing unroll by 8 or even 4 to see how they do if you haven't already.